File: src/lib/api/export.ts
/**
 * Export & Import API
 *
 * Operations for exporting manuscripts and importing content.
 *
 * @module
 */

import { invoke } from '@tauri-apps/api/core';

import type { ImportResult, Scene } from './types';

/**
 * API for export operations.
 */
export const exportApi = {
	markdown: (options?: {
		chapterIds?: string[];
		sceneSeparator?: string;
		includeTitles?: boolean;
	}) =>
		invoke<string>('export_markdown', {
			chapterIds: options?.chapterIds,
			sceneSeparator: options?.sceneSeparator,
			includeTitles: options?.includeTitles,
		}),

	plainText: (options?: { chapterIds?: string[]; sceneSeparator?: string }) =>
		invoke<string>('export_plain_text', {
			chapterIds: options?.chapterIds,
			sceneSeparator: options?.sceneSeparator,
		}),

	jsonBackup: () => invoke<string>('export_json_backup'),

	outline: () => invoke<string>('export_outline'),

	bible: () => invoke<string>('export_bible'),

	timeline: () => invoke<string>('export_timeline'),

	/** Export bible entries as CSV. */
	exportBibleCsv: () => invoke<string>('export_bible_csv'),

	/** Export timeline (events + scenes) as CSV. */
	exportTimelineCsv: () => invoke<string>('export_timeline_csv'),

	/** Export review grid data as CSV. */
	exportReviewGridCsv: () => invoke<string>('export_review_grid_csv'),

	/** Export word count stats as CSV. */
	exportStatsCsv: () => invoke<string>('export_stats_csv'),
};

/**
 * API for import operations.
 */
export const importApi = {
	markdownAsScene: (chapterId: string, title: string, content: string) =>
		invoke<Scene>('import_markdown_as_scene', { chapterId, title, content }),

	markdownStructured: (content: string) =>
		invoke<ImportResult>('import_markdown_structured', { content }),

	textAsScene: (chapterId: string, title: string, content: string) =>
		invoke<Scene>('import_text_as_scene', { chapterId, title, content }),

	/** Import a JSON backup file to restore project data. Creates auto-backup first. */
	jsonBackup: (content: string) => invoke<void>('import_json_backup', { content }),
};

File: src-tauri/src/commands/export.rs
use crate::AppState;
use tauri::State;

#[tauri::command]
pub fn export_markdown(
    chapter_ids: Option<Vec<String>>,
    scene_separator: Option<String>,
    include_titles: Option<bool>,
    state: State<AppState>,
) -> Result<String, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.export_markdown_with_options(
        chapter_ids.as_deref(),
        scene_separator.as_deref(),
        include_titles.unwrap_or(true),
    )
}

#[tauri::command]
pub fn export_plain_text(
    chapter_ids: Option<Vec<String>>,
    scene_separator: Option<String>,
    state: State<AppState>,
) -> Result<String, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.export_plain_text_with_options(chapter_ids.as_deref(), scene_separator.as_deref())
}

#[tauri::command]
pub fn export_json_backup(state: State<AppState>) -> Result<String, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.export_json_backup()
}

#[tauri::command]
pub fn export_outline(state: State<AppState>) -> Result<String, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.export_outline()
}

#[tauri::command]
pub fn export_bible(state: State<AppState>) -> Result<String, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.export_bible()
}

#[tauri::command]
pub fn export_timeline(state: State<AppState>) -> Result<String, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.export_timeline()
}
